### PR TITLE
Add Generic OAuth2 Operator for Grafana

### PR DESCRIPTION
**Pull request by conzetti**
TL;DR:
=====
- Add operator for `generic_oauth` (non-biased OAuth2 configuration)
- Add operator for `root_url` (requisite for OAuth2 redirect)

Detail:
======
These operators allow a deployment to have a generic OAuth2 provider
added as an authentication mechanism for Grafana. This includes all
values that are presented in the [spec](https://github.com/bosh-prometheus/prometheus-boshrelease/blob/master/jobs/grafana/spec#L267-L296).

All variables are configurable, with the exception of the following:
- `grafana.auth.generic_oauth.enabled`
- `grafana.auth.generic_oauth.allow_sign_up`

The `enabled` attribute is our intended purpose of adding this
operator, so that will always be set to `true`. The `allow_sign_up`
attribute provides the ability to allow or deny the creation of users
in Grafana that are provided by the OAuth2 provider. This is detailed
further, [here](http://docs.grafana.org/installation/configuration/#allow-sign-up).
Without this enabled, users will not be able to login.

A generic OAuth2 operator was provided for UAA in
328da574783de6abb078763abffca50a6a978d83, but this is heavily biased
towards using UAA, specifically. For using alternate backends (in my
case, ADFS), a generic operator is required. Additionally, this did
not present the option to rewrite `root_url`.

If opting to use the `enable-grafana-generic-oauth` operator, the
`root_url` may need to be modified to match what the OAuth2 provider
is expecting (in many production use-cases), and also necessary to
make the redirect (call-back) _from_ the OAuth2 provider to the client
functional. The `enable-root-url` operator is intended for this
use-case. Further reading is available, [here](http://docs.grafana.org/installation/configuration/#root-url).

NOTE:
====
These operators were separated to enable the `enable-root-url`
operator to be used without any OAuth2-specific configuration (useful
for proxying scenarios).

### CODE REVIEW

**Comment by frodenas** [comment]
Thanks @conzetti!

**Comment by conzetti** [comment]
Likewise, @frodenas!. 

@robplahn I used the separate ops-file for `root_url` to cover the use-case where someone may want to use a proxy in front of grafana, but doesn't necessarily want to use OAuth2 for AuthN. I had combined them at one point, but talking it over with my peers made me think that it has greater value being a separate ops-file.

**Comment by robplahn** [comment]
@conzetti Thanks, that makes sense. We are exactly that use case. We are using the PCF-Prometheus Concourse pipeline for deployment which sets up an NGINX proxy in front of Grafana.  So this granfana_root_url was required. 
https://github.com/pivotal-cf/pcf-prometheus-pipeline.
